javascript/ql/src/semmle/javascript/frameworks/Electron.qll

```diff
@@ -6,31 +6,466 @@ module Electron {
    */
   class WebPreferences extends DataFlow::ObjectLiteralNode {
     WebPreferences() {
-      exists(BrowserObject bo | this = bo.getOptionArgument(0, "webPreferences").getALocalSource())
+      exists(NewBrowserObject bo | this = bo.getWebPreferences().getALocalSource())
     }
   }
   
+  /**
+   * a data flow node that is an Electron `BrowserView` or `BrowserWindow`
+   */
+  abstract class BrowserObject extends DataFlow::TrackedNode {
+  }
+  
   /**
    * A data flow node that creates a new `BrowserWindow` or `BrowserView`.
    */
-  private abstract class BrowserObject extends DataFlow::NewNode {
+  abstract class NewBrowserObject extends BrowserObject {
+    NewBrowserObject() {
+      this instanceof DataFlow::NewNode
+    }
+    
+    DataFlow::Node getWebPreferences() {
+      result = this.(DataFlow::NewNode).getOptionArgument(0, "webPreferences")
+    }
   }
   
   /**
    * A data flow node that creates a new `BrowserWindow`.
    */
-  class BrowserWindow extends BrowserObject {
+  class BrowserWindow extends NewBrowserObject {
     BrowserWindow() {
       this = DataFlow::moduleMember("electron", "BrowserWindow").getAnInstantiation()
     }
   }
   
+  /**
+   * A data flow node with a TypeScript type indicating it is an Electron `BrowserWindow`
+   */
+  class TypedBrowserWindow extends BrowserObject {
+    TypedBrowserWindow() {
+      this.asExpr().getType().toString() = "BrowserWindow"
```

Comment:
Is this type name sufficiently unique? Also, I'd like to avoid the use of `toString` (which is mainly for debugging).

Reply:
I'll take a deeper look at the TypeScript libraries on Monday and see about adding a namespace check and using something other than `toString`.

Reply:
`Type.hasUnderlyingType()` is supposed to cover this use case (recognizing expressions based on a type name), but it hasn't been used much in the wild yet. If it doesn't work, please talk to me and we'll fix it or make it work somehow.